http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala

```diff
@@ -677,18 +689,22 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
     ctx <- as.ask
     listOfPermissions <- authProvider.getActions(app.samResourceId, userInfo)
 
+    hasWorkspacePermission <- authProvider.isUserWorkspaceReader(WorkspaceResourceSamResourceId(workspaceId), userInfo)
```

Comment:
kind of a general comment, but would it make sense to fail faster when the user doesn't have workspace permission, have that be the first check, error if user doesn't so that we aren't doing the sam checks/other logic unless we have to?

Reply:
That is a very good point, let me do another pass at this and switch things around a bit (hopefully that won't change the unit tests, which were the most annoying things to set up here)